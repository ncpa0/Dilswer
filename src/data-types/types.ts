import type { InstanceOf, SimpleDataType } from "@DataTypes/data-types";
import {
  AllOf,
  ArrayOf,
  Custom,
  Dict,
  Enum,
  EnumMember,
  Literal,
  OneOf,
  RecordOf,
  SetOf,
} from "@DataTypes/data-types";

export {
  AllOf,
  ArrayOf,
  Custom,
  Dict,
  Enum,
  EnumMember,
  Literal,
  OneOf,
  RecordOf,
  SetOf,
};

export interface RecordOfVisitChild<R> {
  _isRecordOfVisitChild: true;
  propertyName: string;
  required: boolean;
  child: R;
}

export interface DataTypeVisitor<R = any> {
  visit(dataType: Exclude<AnyDataType, RecordOf>, children?: R[]): R;
  visit(dataType: RecordOf, children?: RecordOfVisitChild<R>[]): R;
}

export type BasicTypeNames =
  | "string"
  | "number"
  | "boolean"
  | "unknown"
  | "integer"
  | "symbol"
  | "function"
  | "null"
  | "undefined"
  | "stringnumeral"
  | "stringinteger";

export type BasicDataType = SimpleDataType<BasicTypeNames>;

export type ComplexDataType =
  | ArrayOf
  | RecordOf
  | Dict
  | SetOf
  | OneOf
  | AllOf
  | Literal
  | Enum
  | EnumMember
  | InstanceOf
  | Custom;

export type DataTypeKind =
  | "simple"
  | "array"
  | "record"
  | "dictionary"
  | "set"
  | "union"
  | "intersection"
  | "literal"
  | "enumUnion"
  | "enumMember"
  | "instanceOf"
  | "custom";

export type AnyDataType = BasicDataType | ComplexDataType;

export type AllDataTypes = AnyDataType;

export type FieldDescriptor = {
  readonly required?: boolean;
  readonly type: AnyDataType;
};

export interface RecordTypeSchema {
  readonly [key: string]: FieldDescriptor | AnyDataType;
}

export type TypeMetadata<T extends Record<any, any> = Record<any, any>> = {
  description?: string;
  title?: string;
  format?: string;
  extra?: T;
};
