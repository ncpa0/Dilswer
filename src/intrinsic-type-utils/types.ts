import type {
  AllDataTypes,
  BasicDataType,
  FieldDescriptor,
  OneOf,
  RecordOf,
  RecordTypeSchema,
} from "@DataTypes/types";

type GetType<T extends AllDataTypes | FieldDescriptor> =
  T extends FieldDescriptor ? T["type"] : T;

export type RequiredRecord<R extends RecordTypeSchema> = RecordOf<{
  [K in keyof R]: { required: true; type: GetType<R[K]> };
}>;

export type PartialRecord<R extends RecordTypeSchema> = RecordOf<{
  [K in keyof R]: { required: false; type: GetType<R[K]> };
}>;

export type OmitRecord<
  R extends RecordTypeSchema,
  OK extends keyof R
> = RecordOf<{
  [K in keyof R as K extends OK ? never : K]: R[K];
}>;

export type PickRecord<
  R extends RecordTypeSchema,
  PK extends keyof R
> = RecordOf<{
  [K in keyof R as K extends PK ? K : never]: R[K];
}>;

export type SumRecord<
  R1 extends RecordTypeSchema,
  R2 extends RecordTypeSchema
> = RecordOf<
  {
    [K in keyof R1 as K extends keyof R2 ? never : K]: R1[K];
  } & R2
>;

export type ExcludeOneOf<U extends OneOf, E extends BasicDataType> = OneOf<
  Array<Exclude<U["oneOf"][number], E>>
>;
