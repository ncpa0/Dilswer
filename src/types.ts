import type { BasicDataTypes, dataTypeSymbol } from "./type-def-helpers";
import type { ValueOf } from "./type-utils";

export type ArrayOf<DT extends AllDataTypes[] = any[]> = {
  readonly [dataTypeSymbol]: true;
  readonly arrayOf: DT;
};

export type RecordOf<TS extends RecordTypeSchema = RecordTypeSchema> = {
  readonly [dataTypeSymbol]: true;
  readonly recordOf: TS;
};

export type SetOf<DT extends AllDataTypes[] = any[]> = {
  readonly [dataTypeSymbol]: true;
  readonly setOf: DT;
};

export type OneOf<DT extends AllDataTypes[] = any[]> = {
  readonly [dataTypeSymbol]: true;
  readonly oneOf: DT;
};

export type Literal<
  V extends string | number | boolean = string | number | boolean
> = {
  readonly [dataTypeSymbol]: true;
  readonly literal: V;
};

export type Enum<E = any> = {
  readonly [dataTypeSymbol]: true;
  readonly enumInstance: E;
};

export type EnumMember<M = any> = {
  readonly [dataTypeSymbol]: true;
  readonly enumMember: M;
};

export type BasicDataType = ValueOf<typeof BasicDataTypes>;

export type ComplexDataType =
  | ArrayOf
  | RecordOf
  | SetOf
  | OneOf
  | Literal
  | Enum
  | EnumMember;

export type AllDataTypes = BasicDataType | ComplexDataType;

export type FieldDescriptor = {
  readonly required?: boolean;
  readonly type: AllDataTypes;
};

export interface RecordTypeSchema {
  readonly [key: string]: FieldDescriptor | AllDataTypes;
}
