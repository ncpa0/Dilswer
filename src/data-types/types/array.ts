import { BaseType } from "@DataTypes/data-types";
import { getStandardSchemaProps } from "@DataTypes/generate-standard-schema";
import type { ParseDataType, ReWrap } from "@DataTypes/type-utils";
import type { AnyDataType, TypeVisitor } from "@DataTypes/types";
import { Path } from "@Validation/path";
import { ValidationError } from "@Validation/validation-error/validation-error";
import type { StandardSchemaV1 } from "standard-schema";
import { UnionType } from "./union";

export class ArrayType<DT extends AnyDataType[] = any[]> extends BaseType {
  readonly kind = "array";
  private readonly union;

  constructor(public readonly arrayOf: DT) {
    super();
    this.union = new UnionType(this.arrayOf);
    Object.freeze(this.arrayOf);
    Object.freeze(this);
  }

  /** @internal */
  _acceptVisitor<R>(visitor: TypeVisitor<R>): R {
    const children: R[] = [];

    for (let i = 0; i < this.arrayOf.length; i++) {
      children.push(this.arrayOf[i]._acceptVisitor(visitor));
    }

    return visitor.visit(this, children);
  }

  get ["~standard"](): StandardSchemaV1.Props<
    any,
    Array<ReWrap<ParseDataType<DT[number]>>>
  > {
    return getStandardSchemaProps(this);
  }

  ["~validate"](path: Path, value: any): void {
    if (!Array.isArray(value)) throw new ValidationError(path, this, value);

    for (let index = 0; index < value.length; index++) {
      this.union["~validate"](path.concat(index), value[index]);
    }
  }

  ["~matches"](value: any): boolean {
    if (!Array.isArray(value)) return false;

    for (let index = 0; index < value.length; index++) {
      const match = this.union["~matches"](value[index]);
      if (!match) return false;
    }
    return true;
  }
}
