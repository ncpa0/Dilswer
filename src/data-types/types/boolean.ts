import { BaseType } from "@DataTypes/data-types";
import { getStandardSchemaProps } from "@DataTypes/generate-standard-schema";
import type { TypeVisitor } from "@DataTypes/types";
import { Path } from "@Validation/path";
import { ValidationError } from "@Validation/validation-error/validation-error";
import type { StandardSchemaV1 } from "standard-schema";

export class BooleanType extends BaseType {
  readonly kind = "simple";
  public readonly simpleType: "boolean" = "boolean";

  constructor() {
    super();
    Object.freeze(this);
  }

  /** @internal */
  _acceptVisitor<R>(visitor: TypeVisitor<R>): R {
    return visitor.visit(this);
  }

  get ["~standard"](): StandardSchemaV1.Props<any, boolean> {
    return getStandardSchemaProps(this);
  }

  ["~validate"](path: Path, value: any): void {
    if (typeof value !== "boolean") {
      throw new ValidationError(path, this, value);
    }
  }

  ["~matches"](value: any): boolean {
    return typeof value === "boolean";
  }
}
