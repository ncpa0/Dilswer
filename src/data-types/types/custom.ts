import { BaseType } from "@DataTypes/data-types";
import { getStandardSchemaProps } from "@DataTypes/generate-standard-schema";
import type { GetFnAssertType } from "@DataTypes/type-utils";
import type { TypeVisitor } from "@DataTypes/types";
import { Path } from "@Validation/path";
import { ValidationError } from "@Validation/validation-error/validation-error";
import type { StandardSchemaV1 } from "standard-schema";

export class CustomType<
  VF extends (v: any) => v is any = (v: any) => v is unknown,
> extends BaseType {
  readonly kind = "custom";

  constructor(public readonly custom: VF) {
    super();
    Object.freeze(this);
  }

  /** @internal */
  _acceptVisitor<R>(visitor: TypeVisitor<R>): R {
    return visitor.visit(this);
  }

  get ["~standard"](): StandardSchemaV1.Props<any, GetFnAssertType<VF>> {
    return getStandardSchemaProps(this);
  }

  ["~validate"](path: Path, value: any): void {
    if (!this.custom(value)) throw new ValidationError(path, this, value);
  }

  ["~matches"](value: any): boolean {
    return this.custom(value);
  }
}
