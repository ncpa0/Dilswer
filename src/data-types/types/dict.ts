import { BaseType } from "@DataTypes/data-types";
import { getStandardSchemaProps } from "@DataTypes/generate-standard-schema";
import type { ParseDataType, ReWrap } from "@DataTypes/type-utils";
import type { AnyType, TypeVisitor } from "@DataTypes/types";
import { Path } from "@Validation/path";
import { ValidationError } from "@Validation/validation-error/validation-error";
import type { StandardSchemaV1 } from "standard-schema";
import { UnionType } from "./union";

export class DictType<DT extends AnyType[] = any[]> extends BaseType {
  readonly kind = "dictionary";
  private readonly union;

  constructor(public readonly dict: DT) {
    super();
    this.union = new UnionType(this.dict);
    Object.freeze(this.dict);
    Object.freeze(this);
  }

  /** @internal */
  _acceptVisitor<R>(visitor: TypeVisitor<R>): R {
    const children: R[] = [];

    for (let i = 0; i < this.dict.length; i++) {
      children.push(this.dict[i]._acceptVisitor(visitor));
    }

    return visitor.visit(this, children);
  }

  get ["~standard"](): StandardSchemaV1.Props<
    any,
    Record<string | number, ReWrap<ParseDataType<DT[number]>>>
  > {
    return getStandardSchemaProps(this);
  }

  ["~validate"](path: Path, value: any): void {
    if (
      typeof value !== "object"
      || value === null
      || Array.isArray(value)
    ) {
      throw new ValidationError(path, this, value);
    }

    const keys = Object.keys(value);

    for (let index = 0; index < keys.length; index++) {
      const key = keys[index];
      this.union["~validate"](path.concat(key), value[key]);
    }
  }

  ["~matches"](value: any): boolean {
    if (
      typeof value !== "object"
      || value === null
      || Array.isArray(value)
    ) {
      return false;
    }

    const keys = Object.keys(value);

    for (let index = 0; index < keys.length; index++) {
      const key = keys[index];
      const match = this.union["~matches"](value[key]);
      if (!match) return false;
    }
    return true;
  }
}
