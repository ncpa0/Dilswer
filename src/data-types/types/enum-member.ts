import { BaseType, MetadataSymbol } from "@DataTypes/data-types";
import { getStandardSchemaProps } from "@DataTypes/generate-standard-schema";
import type { TypeMetadata, TypeVisitor } from "@DataTypes/types";
import { Path } from "@Validation/path";
import { ValidationError } from "@Validation/validation-error/validation-error";
import type { StandardSchemaV1 } from "standard-schema";

export class EnumMemberType<DT = any> extends BaseType {
  /** @internal */
  static getOriginalMetadata(dt: EnumMemberType) {
    return dt[MetadataSymbol];
  }

  protected [MetadataSymbol]: TypeMetadata & {
    memberName?: `${string}`;
    enumName?: string;
  } = {};

  readonly kind = "enumMember";
  constructor(public readonly enumMember: DT) {
    super();
    Object.freeze(this);
  }

  /**
   * Sets the metadata for the enum name. This is used for
   * generating appropriate TypeScript declarations (via
   * `toTsType()`).
   */
  setEnumName<T extends EnumMemberType>(this: T, name: string): T {
    this[MetadataSymbol].enumName = name;
    return this;
  }

  /**
   * Sets the metadata for the enum member name. This is used for
   * generating appropriate TypeScript declarations (via
   * `toTsType()`).
   */
  setMemberName<T extends EnumMemberType>(this: T, name: `${string}`): T {
    this[MetadataSymbol].memberName = name;
    return this;
  }

  /** @internal */
  _acceptVisitor<R>(visitor: TypeVisitor<R>): R {
    return visitor.visit(this);
  }

  get ["~standard"](): StandardSchemaV1.Props<any, DT> {
    return getStandardSchemaProps(this);
  }

  ["~validate"](path: Path, value: any): void {
    if (this.enumMember !== value) {
      throw new ValidationError(path, this, value);
    }
  }

  ["~matches"](value: any): boolean {
    return this.enumMember === value;
  }
}
