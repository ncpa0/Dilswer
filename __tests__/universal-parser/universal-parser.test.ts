import { RecordVisitChild } from "@DataTypes/types";
import {
  AnyDataType,
  DataTypeVisitor,
  parseWith,
  RecordType,
  Type,
} from "../../src/index";

describe("parseWith", () => {
  it("should parse a type using the provided visitor interface", () => {
    type TestParsedFormat = {
      name: string;
      simpleType?: string;
      children: TestParsedFormat[] | RecordVisitChild<TestParsedFormat>[];
    };

    class TestVisitor implements DataTypeVisitor<TestParsedFormat> {
      visit(
        dataType: Exclude<AnyDataType, RecordType>,
        children?: TestParsedFormat[],
      ): TestParsedFormat;
      visit(
        dataType: RecordType,
        children?: RecordVisitChild<TestParsedFormat>[],
      ): TestParsedFormat;
      visit(dataType: AnyDataType, children: any[] = []): TestParsedFormat {
        return {
          name: dataType.kind,
          simpleType: dataType.kind === "simple"
            ? dataType.simpleType
            : undefined,
          children: children.map((c) =>
            c._isRecordOfVisitChild
              ? { ...c.child, propertyName: c.propertyName }
              : c
          ),
        };
      }
    }

    const parsed = parseWith(
      new TestVisitor(),
      Type.Record({
        foo: Type.String,
        bar: Type.Number,
        baz: Type.Record({
          qux: Type.Boolean,
          quux: Type.Array(Type.String),
        }),
      }),
    );

    expect(parsed).toEqual({
      name: "record",
      children: [
        {
          propertyName: "foo",
          name: "simple",
          simpleType: "string",
          children: [],
        },
        {
          propertyName: "bar",
          name: "simple",
          simpleType: "number",
          children: [],
        },
        {
          propertyName: "baz",
          name: "record",
          simpleType: undefined,
          children: [
            {
              propertyName: "qux",
              name: "simple",
              simpleType: "boolean",
              children: [],
            },
            {
              propertyName: "quux",
              name: "array",
              simpleType: undefined,
              children: [
                {
                  name: "simple",
                  simpleType: "string",
                  children: [],
                },
              ],
            },
          ],
        },
      ],
    });
  });

  it("readme example test", () => {
    type Node = {
      typeName: string;
      children?: Node[] | Record<string, Node>;
    };

    const visitor = {
      visit(
        type: AnyDataType,
        children?: Node[] | RecordVisitChild<Node>[],
      ): Node {
        switch (type.kind) {
          case "simple":
            return { typeName: type.simpleType };
          case "record":
            return {
              typeName: "record",
              children: children
                && Object.fromEntries(
                  (children as RecordVisitChild<Node>[]).map(
                    ({ propertyName, child }) => [propertyName, child],
                  ),
                ),
            };
          default:
            return { typeName: type.kind, children: children as Node[] };
        }
      },
    };

    const type = Type.Record({
      foo: Type.String,
      bar: Type.Array(Type.Number),
      baz: Type.OneOf(Type.String, Type.Number),
    });

    const nodeTree = parseWith(visitor, type);

    expect(nodeTree).toMatchSnapshot();
  });
});
