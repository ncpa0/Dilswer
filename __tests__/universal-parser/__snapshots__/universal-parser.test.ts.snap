// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseWith readme example test 1`] = `
{
  "children": {
    "bar": {
      "children": [
        {
          "typeName": "number",
        },
      ],
      "typeName": "array",
    },
    "baz": {
      "children": [
        {
          "typeName": "string",
        },
        {
          "typeName": "number",
        },
      ],
      "typeName": "union",
    },
    "foo": {
      "typeName": "string",
    },
  },
  "typeName": "record",
}
`;
